### PR TITLE
feat: add accuracy measure for mlpack_logistic_regression

### PR DESCRIPTION
**Pull request by rabbabansh**
addresses issue #3552 

I implemented "print_traning_accuracy" in logistic_regression_main.cpp as implemented in random_forest_main.cpp.

@rcurtin can you please review this pr?

### CODE REVIEW

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 1]
```diff
@@ -153,6 +153,9 @@ PARAM_MATRIX_OUT("probabilities", "If test data is specified, this "
 PARAM_DOUBLE_IN("decision_boundary", "Decision boundary for prediction; if the "
     "logistic function for a point is less than the boundary, the class is "
     "taken to be 0; otherwise, the class is 1.", "d", 0.5);
+PARAM_FLAG("print_training_accuracy", "If set, then the accuracy of the model "
+    "on the training set will be predicted (verbose must also be specified).",
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:157]:
```suggestion
    "on the training set will be printed (verbose must also be specified).",
```

File: src/mlpack/methods/random_forest/random_forest_main.cpp [hunk 2]
```diff
@@ -24,14 +24,14 @@ using namespace std;
...
-// Short description.
+// Short description
 BINDING_SHORT_DESC(
     "An implementation of the standard random forest algorithm by Leo Breiman "
     "for classification.  Given labeled data, a random forest can be trained "
     "and saved for future use; or, a pre-trained random forest can be used for "
     "classification.");
 
-// Long description.
+// Long description
```
Comment (rcurtin) [inline src/mlpack/methods/random_forest/random_forest_main.cpp:34]: I don't see the reason for these changes.

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 3]
```diff
@@ -327,6 +334,22 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
...
+      model->Classify(regressors, predictions);
+
+      const size_t correct = arma::accu(predictions == responses);
+
+      Log::Info << correct << " of " << responses.n_elem << " correct on training"
+          << " set (" << (double(correct) / double(responses.n_elem) * 100) << ")."
+          << endl;
+      timers.Stop("lr_prediction");
+    }
+  }
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:351]: It seems like there's an extra brace here, and things are indented more than they need to be.

File: HISTORY.md [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
   * Use HTTPS for all auto-downloaded dependencies (#3550).
 
+  * Feat: Add "print_training_accuracy" in logistic_regression_main.md ([#3552](https://github.com/mlpack/mlpack/issues/3552))
```
Comment (rcurtin) [inline HISTORY.md:9]:
```suggestion
  * Add `print_training_accuracy` option to LogisticRegression bindings (#3552).
```

**Review by rcurtin — COMMENTED** [review]
Thanks for working on this @rabbabansh. :+1: I think the changes look good, but can I ask you to go through and make sure that the code matches the [style guide](https://github.com/mlpack/mlpack/wiki/DesignGuidelines#style-guidelines)?  Specifically there are a lot of lines that are longer than 80 characters.  If you can reflow them, I would appreciate it.

Thanks!

**Comment by rabbabansh** [comment]
I have resolved all the suggestions, is there anything more that would make the code better? @rcurtin

**Comment by rabbabansh** [comment]
Can you please review this PR @rcurtin ?

**Comment by rcurtin** [comment]
When I have time, I'll get back to it.  It is on my todo list already.

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 5]
```diff
@@ -182,6 +185,10 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
   ReportIgnoredParam(params, {{ "test", false }}, "predictions");
   ReportIgnoredParam(params, {{ "test", false }}, "probabilities");
 
+  ReportIgnoredParam(params, {{ "training", false }}, "print_training_accuracy");
+  
+  RequireAtLeastOnePassed(params, { "test", "output_model", "print_training_accuracy" }, false, "the trained logistic regression model will not be used or saved");
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:190]:
```suggestion
  RequireAtLeastOnePassed(params,
      { "test", "output_model", "print_training_accuracy" }, false,
      "the trained logistic regression model will not be used or saved");
```
(line length fix)

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 6]
```diff
@@ -101,7 +101,7 @@ BINDING_EXAMPLE(
     PRINT_MODEL("lr_model") + "', the following command may be used:"
     "\n\n" +
     PRINT_CALL("logistic_regression", "training", "data", "labels", "labels",
-        "lambda", 0.1, "output_model", "lr_model") +
+        "lambda", 0.1, "output_model", "lr_model", "print_training_accuracy", true) +
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:104]:
```suggestion
        "lambda", 0.1, "output_model", "lr_model", "print_training_accuracy",
        true) +
```
(line length fix)

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 7]
```diff
@@ -327,6 +334,22 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
...
+
+    const size_t correct = arma::accu(predictions == responses);
+
+    Log::Info << correct << " of " << responses.n_elem << " correct on training"
+        << " set (" << (double(correct) / double(responses.n_elem) * 100) << ")."
+        << endl;
+    timers.Stop("lr_prediction");
+  }
+  
+
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:352]:
```suggestion
```
No need for an extra line there. :+1:

**Review by rcurtin — COMMENTED** [review]
Looks good.  I should have mentioned this last time, but would you mind adding a very simple test in `src/mlpack/tests/main_tests/logistic_regression_test.cpp`?  It's not reasonable to check the exact output, so I think a reasonable test could just run the binding with `print_training_accuracy` and make sure it doesn't crash.  (This will help us ensure that in the future, `Classify()` isn't being called wrong or anything in that block.)

Thanks for the contribution, it is much appreciated. :+1:

**Comment by rabbabansh** [comment]
@rcurtin added a test for print_training_accuracy, please have a look and let me know if I need to make any more changes.

**Comment by rabbabansh** [comment]
hey @rcurtin, any updates?

File: src/mlpack/tests/main_tests/logistic_regression_test.cpp [hunk 8]
```diff
@@ -626,4 +626,25 @@ TEST_CASE_METHOD(LogisticRegressionTestFixture, "LRDecisionBoundaryTest",
...
+   * does not crash.
+   */
+  TEST_CASE_METHOD(LogisticRegressionTestFixture, "LRPrintTrainingAccuracyTest",
+                  "[LogisticRegressionMainTest][BindingTests]")
+  {
+    constexpr int N = 100;
+    constexpr int D = 5;
+
+    arma::mat trainX = arma::randu<arma::mat>(D, N);
+    arma::Row<size_t> trainY(N, arma::fill::randu);
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/logistic_regression_test.cpp:641]:
```suggestion
    arma::Row<size_t> trainY = arma::randi<arma::Row<size_t>>(N,
        arma::distr_param(0, 1));
```
Technically for type size_t I believe randu will result in a vector of all zeros.  You could instead do this.  (I think the test is fine either way.)

**Review by rcurtin — COMMENTED** [review]
Thanks, looks good.  Let's make sure it passes the CI checks and then let's merge it. :+1:

**Comment by rcurtin** [comment]
Can you merge master into this branch?  I don't think I have permission to push to your master branch.

**Comment by rabbabansh** [comment]
Done!

**Review by rcurtin — COMMENTED** [review]
The tests fail to compile because you put the new test inside of another test; can you fix that?  Then I think this is good to merge.  I'd suggest compiling and testing locally, instead of depending on CI to find the issue. :+1:

**Comment by rabbabansh** [comment]
Fixed the test, anything else? @rcurtin

File: src/mlpack/tests/main_tests/logistic_regression_test.cpp [hunk 9]
```diff
@@ -627,3 +627,25 @@ TEST_CASE_METHOD(LogisticRegressionTestFixture, "LRDecisionBoundaryTest",
   // Check that the output changed when the decision boundary moved.
   REQUIRE(arma::accu(output1 != output2) > 0);
 }
+
+/**
+  * Check that running the binding with print_training_accuracy set to true
+  * does not crash.
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/logistic_regression_test.cpp:633]:
```suggestion
 * Check that running the binding with print_training_accuracy set to true
 * does not crash.
```
(Really trivial spacing fix)

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 10]
```diff
@@ -327,6 +337,21 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
...
+    arma::Row<size_t> predictions;
+    model->Classify(regressors, predictions);
+
+    const size_t correct = arma::accu(predictions == responses);
+
+    Log::Info << correct << " of " << responses.n_elem << " correct on training"
+        << " set (" << (double(correct) / double(responses.n_elem) * 100) << ")."
+        << endl;
+    timers.Stop("lr_prediction");
+  }
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:353]: I realized while reviewing this today that this block should actually be inside the `if (params.Has("training"))` block above: if the user does not pass training data, but does pass `print_training_accuracy`, the option will not be ignored: we will try to print training accuracy anyway.

**Review by rcurtin — COMMENTED** [review]
Sorry, found just one more thing!  Everything else looks good though; as soon as we fix this and the tests pass I think it's ready to merge. :+1:

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp (hunk 10 above)
Comment (rabbabansh) [reply src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:353]: I added the `if (params.Has("training"))` to the block, is that all?

File: src/mlpack/methods/logistic_regression/logistic_regression_main.cpp [hunk 11]
```diff
@@ -299,6 +309,21 @@ void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
...
+      arma::Row<size_t> predictions;
+      model->Classify(regressors, predictions);
+
+      const size_t correct = arma::accu(predictions == responses);
+
+      Log::Info << correct << " of " << responses.n_elem << " correct on training"
+          << " set (" << (double(correct) / double(responses.n_elem) * 100) << ")."
+          << endl;
+      timers.Stop("lr_prediction");
+    }
```
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression_main.cpp:325]: Thanks for moving the block, but we can't print accuracy before we train the model.  Please move it down below the training code (to the end of this `if` block).

**Review by rcurtin — APPROVED** [review]
Thanks for the contribution!  You should merge master into this branch; there will be a HISTORY.md conflict to resolve (but it should be easy).

**Comment by rabbabansh** [comment]
Is that all?

**Comment by rabbabansh** [comment]
Fixed the history.md release change-log issue!

**Comment by rabbabansh** [comment]
any updates? @rcurtin

**Comment by rcurtin** [comment]
I was out of town over the weekend.  Thanks for handling the issues.

**Comment by rabbabansh** [comment]
Kudos!
